File: src/dexter/tools/finance/segments.py
from langchain.tools import tool
from typing import Literal
from pydantic import BaseModel, Field
from dexter.tools.finance.api import call_api

####################################
# Tools
####################################

class SegmentedRevenuesInput(BaseModel):
    """Input for the get_segmented_revenues tool."""

    ticker: str = Field(description="The stock ticker symbol to fetch segmented revenues for. For example, 'AAPL' for Apple.")
    period: Literal["annual", "quarterly"] = Field(description="The reporting period for the segmented revenues. 'annual' for yearly, 'quarterly' for quarterly.")
    limit: int = Field(default=10, description="The number of past periods to retrieve.")


@tool(args_schema=SegmentedRevenuesInput)
def get_segmented_revenues(
    ticker: str,
    period: Literal["annual", "quarterly"],
    limit: int = 10,
) -> dict:
    """Provides a detailed breakdown of a company's revenue by operating segments, such as products, services, or geographic regions. Useful for analyzing the composition of a company's revenue."""
    params = {
        "ticker": ticker,
        "period": period,
        "limit": limit
    }
    data = call_api("/financials/segmented-revenues/", params)
    return data.get("segmented_revenues", {})

File: src/tools/finance/segments.ts
import { DynamicStructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import { fmp } from './api.js';
import { formatToolResult } from '../types.js';

export const getSegmentedRevenues = new DynamicStructuredTool({
  name: 'get_segmented_revenues',
  description: `Provides a breakdown of a company's revenue by product segments and geographic regions via FMP. Useful for analyzing revenue composition and geographic diversification.`,
  schema: z.object({
    ticker: z
      .string()
      .describe("The stock ticker symbol. For example, 'AAPL' for Apple."),
    segment_type: z
      .enum(['product', 'geographic'])
      .default('product')
      .describe("Type of segmentation: 'product' for business lines, 'geographic' for regions."),
  }),
  func: async (input) => {
    const ticker = input.ticker.trim().toUpperCase();
    const endpoint =
      input.segment_type === 'geographic'
        ? `/revenue-geographic-segmentation/${ticker}`
        : `/revenue-product-segmentation/${ticker}`;
    const { data, url } = await fmp.get(endpoint, { structure: 'flat' });
    return formatToolResult(Array.isArray(data) ? data : [], [url]);
  },
});
